File: java/circular-buffer/src/main/java/CircularBuffer.java
public class CircularBuffer<T> {
	T[] buffer;
	int headIndex;
	int tailIndex;
	int length;

	@SuppressWarnings("unchecked")
	CircularBuffer(int size) {
		buffer = (T[]) new Object[size];

		clear();
	}

	T read() throws BufferIOException {
		if (isEmpty()) {
			throw new BufferIOException("Tried to read from empty buffer");
		}

		T result = buffer[headIndex];
		headIndex = (headIndex + 1) % buffer.length;
		length--;
		return result;
	}

	void write(T element) throws BufferIOException {
		if (isFull()) {
			throw new BufferIOException("Tried to write to full buffer");
		}

		buffer[tailIndex] = element;
		tailIndex = (tailIndex + 1) % buffer.length;
		length++;
	}

	void overwrite(T element) throws BufferIOException {
		if (isFull()) {
			read();
		}
		write(element);
	}

	void clear() {
		headIndex = 0;
		tailIndex = 0;
		length = 0;
	}

	boolean isEmpty() {
		return length == 0;
	}

	boolean isFull() {
		return length > 0 && headIndex == tailIndex;
	}
}

File: python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, capacity):
        self.buffer = [None] * capacity
        self.clear()

    def read(self):
        if self.is_empty():
            raise BufferEmptyException('Buffer is empty!')

        result = self.buffer[self.head_index]
        self.head_index = (self.head_index + 1) % len(self.buffer)
        self.length -= 1
        return result

    def write(self, element):
        if self.is_full():
            raise BufferFullException('Buffer is full!')

        self.buffer[self.tail_index] = element
        self.tail_index = (self.tail_index + 1) % len(self.buffer)
        self.length += 1

    def overwrite(self, element):
        if self.is_full():
            self.read()
        self.write(element)

    def clear(self):
        self.head_index = 0
        self.tail_index = 0
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def is_full(self):
        return self.length > 0 and self.head_index == self.tail_index
